Navigation Enhancements x2

nahwneeth opened the pull request:
Continuing work done in #670

# Scoped view models
- All the view models that were declared in the activity are now scoped to the routes in which they are used. View models used by most of the routes (like SiteViewModel, AccountViewModel & AccountSettingsViewModel) are left untouched.
- This prevents sharing view models that lead to stale data & reloading the data when navigating back.
- Making the LaunchedEffect run only once per route. This helps ensure that the content isn't reset with new data when moving back from one route to another. Things like scroll position & filters are also retained.

Closes #704
Closes #402 
Closes #774

# Declarative navigation
- Moved all navigate calls to the MainActivity making it easy to see all the possible navigations (the full graph).
- Dependencies of a route are made explicit and compile safe.
- It's now possible to pass callbacks from one route to another. This is what made scoping view models a possibility in the first place. Here's a snippet from the code where replying to a post/comment should append the comment on success:
  ```kotlin
  navController.toCommentReply.navigate(
      CommentReplyDependencies(
          ReplyItem.PostItem(pv),
          isModerator = isModerator(
              pv.creator,
              postRes.data.moderators,
          ),
          onCommentReply = postViewModel::appendComment,
      ),
  )
  ```
- Any nav controllers that were passed to the view model have been brought out during this refactor to make it easier to test view models.

# Bottom navigation bar
- The bottom app bar is now only present in the home activity which uses a second nav host for the bottom navigation.
- Clicking on Inbox, Saved, Profile tabs & create post button now opens the login activity if user is not logged in.
- The inbox deeplink has been taken care of by redirecting to the home activity with the inbox tab as default. The default is reset to feed if the user is not logged in.

Closes #714
Closes #722 (Closes #719 similar)
Partially addresses #346 

# Screen transitions
- Added [Accompanist navigation animation](https://google.github.io/accompanist/navigation-animation/) as dependency.
- All the main navigations have a slide in and out transition by default.
- Removed the jerky fade in & out transition which was being cased by background set as transparent.

Closes #300 

# Other changes
- The back button auto hides/shows based on the nav stack state. So opening screens using deep links or the nested screens in the `HomeActivity` don't have a back button.
- Community & profiles open as separate screen instead of within the "search" tab.
- Fixed horizontal tab swipe in the profile activity which caused by the view not being big enough.

Closes #789 

I've tried not to modify any logic other than navigation. Let me know if anything is not working as expected.

dessalines commented:
Whoa this is huge! I pry won't be able to test / go over this for a few days, because it has so many changes.

> All the view models that were declared in the activity are now scoped to the routes in which they are used. View models used by most of the routes (like SiteViewModel, AccountViewModel & AccountSettingsViewModel) are left untouched.

This is my main concern. ViewModels are supposed to exist outside of activities, so that they can persist without them. Otherwise things like, going to a post, then clicking someone's profile, then clicking the back button, *will reload the post*, rather than just going to place you were at in the post before. Does this PR mean that essentially all back button, and viewing of old routes, essentially has to get loaded from scratch?

Not only that, but thing like comment edits, that initialize data into the view models, would have to start with fresh data.

MV-GH commented:
I have been testing this build since yesterday, and i haven't encountered any issues yet
> so that they can persist without them. Otherwise things like, going to a post, then clicking someone's profile, then clicking the back button, will reload the post, rather than just going to place you were at in the post before.

It behaves like before.

In accidentally (or I assume so) is that it removes the navController from many things. Which prevents now the constant recompositions of the bottamappbar . But sadly the scrolling still feels janky. So there are more underlying problems

nahwneeth commented:
The reloading thing has been taken care of. I've created a new composable called `InitializeRoute` that is essentially `LaunchedEffect` that runs only once per route.

```kotlin
@Composable
fun InitializeRoute(initializationBlock: suspend CoroutineScope.() -> Unit) {
    // rememberSaveable persists this information across navigations
    var initialized by rememberSaveable { mutableStateOf(false) }

    // LaunchedEffect runs everytime we navigate. Therefore we need a guard.
    if (!initialized) {
        LaunchedEffect(Unit) {
            this.initializationBlock()

            // The effects might get cancelled sometimes and the initialization wouldn't have happened.
            // Therefore this has to be inside the effect and after the initialization block.
            initialized = true
        }
    }
}
```

In the post activity we do this:
```kotlin
val postViewModel: PostViewModel = viewModel()
InitializeRoute {
    postViewModel.initialize(id = postArg)
    postViewModel.getData(account)
}
```

The view model gets scoped to the `navBackStackEntry`. So, as long as the back entry exists and not popped from the stack, we'll get the same view model whenever we navigate to that route. So in this case the `postViewModel` is initialized the first time the route is composed and never again. 

Going to a profile and coming back will not reload the route. The view model, the scroll positions and the filters that were applied to the comments will also be retained.

I can add a video demonstrating all the improvements if it helps.

nahwneeth commented:
The commit is huge largely because of making the routes type safe. Most of them are new files listing dependencies and the possible navigations and the extracting the arguments from the route in a typesafe manner. In the existing files, only the `navController.navigate(...)` lines have been removed and replaced with typesafe ones.

Do take your time to test it, no worries. Do let me know if there are any issues.

MV-GH commented on app/src/main/java/com/jerboa/ui/components/home/HomeActivity.kt:
```diff
@@ -84,282 +73,149 @@ fun HomeActivity(
+        }
+    }
+
+    val onSelectTab = { tab: HomeTab ->
+        if (tab.needsLogin() && account == null) {
+            feedNavController.toLogin.navigate()
+        } else {
```
This is a change in behavior though, it will navigate to the login instead of displaying a toast like before. Not sure if this is wanted.

twizmwazin replied:
One of the frequent complaints by new users is that it is difficult to find the option to log in. Redirecting users to log in when they attempt to use a function that requires being logged in seems like a good solution to that problem that many have requested.

MV-GH replied:
Well one of the other suggestions where to add a login button in the more options dialog and/or show login snackbar on initial app use.

It still doesn't solve the root problem as new users still won't know where to switch/log out.

And now is inconsistent with the other buttons that require login(upvotes , ...) that still display display this toast.

Either way, the current dropdown should be made more clear. Maybe we could ve a "Click here to login" text next to anonymous. That only shows on initial startup and diseapears once you click on the anonymous drop down.

nahwneeth replied:
For the bottom nav bar, the nav controller was already available so went ahead with navigating to the login page. For the upvotes, downvotes and reply, the nav controller (or callbacks) have to be passed deeper down the tree. There's a TODO to show the login page along with the toast, so I assume this will be revisited.

https://github.com/dessalines/jerboa/blob/b8e0dfb113e0e4ef40dd539ebbaed5ee3aedf05f/app/src/main/java/com/jerboa/Utils.kt#L116-L119

Agree with the log out thing.

MV-GH commented:
Found some issues though, clicking on @ linking to subs causes a crash

 java.lang.IllegalArgumentException: Navigation destination that matches request NavDeepLinkRequest{ uri=android-app://androidx.navigation/lemmy.world/c/syncforlemmy } cannot be found in the navigation graph d0(0x0) startDestination={a(0x8a77bdc8) route=home?tab={tab}}
 
 <details><summary>

Full stack trace

</summary>

java.lang.IllegalArgumentException: Navigation destination that matches request NavDeepLinkRequest{ uri=android-app://androidx.navigation/lemmy.world/c/syncforlemmy } cannot be found in the navigation graph d0(0x0) startDestination={a(0x8a77bdc8) route=home?tab={tab}}
                                                                                                    	at c3.r.n(Unknown Source:121)
                                                                                                    	at w4.s0.z(Unknown Source:590)
                                                                                                    	at g5.e1.a(Unknown Source:21)
                                                                                                    	at e6.g.onClick(Unknown Source:4)
                                                                                                    	at android.text.method.LinkMovementMethod.onTouchEvent(LinkMovementMethod.java:232)
                                                                                                    	at g6.a.onTouchEvent(Unknown Source:2)
                                                                                                    	at android.widget.TextView.onTouchEvent(TextView.java:10981)
                                                                                                    	at android.view.View.dispatchTouchEvent(View.java:14381)
                                                                                                    	at android.view.ViewGroup.dispatchTransformedTouchEvent(ViewGroup.java:3151)
                                                                                                    	at android.view.ViewGroup.dispatchTouchEvent(ViewGroup.java:2830)
                                                                                                    	at h1.c0.j0(Unknown Source:52)
                                                                                                    	at m.o.j0(Unknown Source:74)
                                                                                                    	at s1.b.H(Unknown Source:35)
                                                                                                    	at h1.a0.S(Unknown Source:90)
                                                                                                    	at h1.a0.R(Unknown Source:66)
                                                                                                    	at m1.e.v(Unknown Source:13)
                                                                                                    	at h1.g.e(Unknown Source:46)
                                                                                                    	at h1.g.e(Unknown Source:73)
                                                                                                    	at h1.g.e(Unknown Source:73)
                                                                                                    	at h1.g.e(Unknown Source:73)
                                                                                                    	at h1.g.e(Unknown Source:73)
                                                                                                    	at h1.h.e(Unknown Source:25)
                                                                                                    	at i.b.c(Unknown Source:26)
                                                                                                    	at d0.y.c(Unknown Source:182)
                                                                                                    	at androidx.compose.ui.platform.AndroidComposeView.D(Unknown Source:77)
                                                                                                    	at androidx.compose.ui.platform.AndroidComposeView.l(Unknown Source:220)
                                                                                                    	at androidx.compose.ui.platform.AndroidComposeView.dispatchTouchEvent(Unknown Source:87)
                                                                                                    	at android.view.ViewGroup.dispatchTransformedTouchEvent(ViewGroup.java:3151)
                                                                                                    	at android.view.ViewGroup.dispatchTouchEvent(ViewGroup.java:2830)
                                                                                                    	at android.view.ViewGroup.dispatchTransformedTouchEvent(ViewGroup.java:3151)
                                                                                                    	at android.view.ViewGroup.dispatchTouchEvent(ViewGroup.java:2830)
                                                                                                    	at android.view.ViewGroup.dispatchTransformedTouchEvent(ViewGroup.java:3151)
                                                                                                    	at android.view.ViewGroup.dispatchTouchEvent(ViewGroup.java:2830)
                                                                                                    	at android.view.ViewGroup.dispatchTransformedTouchEvent(ViewGroup.java:3151)
                                                                                                    	at android.view.ViewGroup.dispatchTouchEvent(ViewGroup.java:2830)
                                                                                                    	at com.android.internal.policy.DecorView.superDispatchTouchEvent(DecorView.java:501)
                                                                                                    	at com.android.internal.policy.PhoneWindow.superDispatchTouchEvent(PhoneWindow.java:1875)
                                                                                                    	at android.app.Activity.dispatchTouchEvent(Activity.java:4225)
                                                                                                    	at com.android.internal.policy.DecorView.dispatchTouchEvent(DecorView.java:459)
                                                                                                    	at android.view.View.dispatchPointerEvent(View.java:14655)
                                                                                                    	at android.view.ViewRootImpl$ViewPostImeInputStage.processPointerEvent(ViewRootImpl.java:6278)
                                                                                                    	at android.view.ViewRootImpl$ViewPostImeInputStage.onProcess(ViewRootImpl.java:6065)
                                                                                                    	at android.view.ViewRootImpl$InputStage.deliver(ViewRootImpl.java:5520)
                                                                                                    	at android.view.ViewRootImpl$InputStage.onDeliverToNext(ViewRootImpl.java:5577)
                                                                                                    	at android.view.ViewRootImpl$InputStage.forward(ViewRootImpl.java:5543)
                                                                                                    	at android.view.ViewRootImpl$AsyncInputStage.forward(ViewRootImpl.java:5695)
                                                                                                    	at android.view.ViewRootImpl$InputStage.apply(ViewRootImpl.java:5551)
                                                                                                    	at android.view.ViewRootImpl$AsyncInputStage.apply(ViewRootImpl.java:5752)
                                                                                                    	at android.view.ViewRootImpl$InputStage.deliver(ViewRootImpl.java:5524)
                                                                                                    	at android.view.ViewRootImpl$InputStage.onDeliverToNext(ViewRootImpl.java:5577)
                                                                                                    	at android.view.ViewRootImpl$InputStage.forward(ViewRootImpl.java:5543)
                                                                                                    	at android.view.ViewRootImpl$InputStage.apply(ViewRootImpl.java:5551)
                                                                                                    	at android.view.ViewRootImpl$InputStage.deliver(ViewRootImpl.java:5524)
                                                                                                    	at android.view.ViewRootImpl.deliverInputEvent(ViewRootImpl.java:8404)
                                                                                                    	at android.view.ViewRootImpl.doProcessInputEvents(ViewRootImpl.java:8355)
                                                                                                    	at android.view.ViewRootImpl.enqueueInputEvent(ViewRootImpl.java:8315)
                                                                                                    	at android.view.ViewRootImpl$WindowInputEventReceiver.onInputEvent(ViewRootImpl.java:8540)
                                                                                                    	at android.view.InputEventReceiver.dispatchInputEvent(InputEventReceiver.java:241)
                                                                                                    	at android.os.MessageQueue.nativePollOnce(Native Method)
                                                                                                    	at android.os.MessageQueue.next(MessageQueue.java:335)
                                                                                                    	at android.os.Looper.loop(Looper.java:193)
2023-06-21 00:29:30.156 18686-18686 MessageQueue-JNI        com.jerboa                           E  	at android.app.ActivityThread.main(ActivityThread.java:8061)
                                                                                                    	at java.lang.reflect.Method.invoke(Native Method)
                                                                                                    	at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:656)
                                                                                                    	at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:967)
</details>

nahwneeth commented:
> Found some issues though, clicking on @ linking to subs causes a crash (community@instance) 
> So does @community.

The last commit fixes this. So try with the latest changes.

MV-GH commented:
> > Found some issues though, clicking on @ linking to subs causes a crash (community@instance)
> > So does @community.
> 
> The last commit fixes this. So try with the latest changes. Thanks.

I did and i just fetched it again and im having the exact same errors.

Am i perhaps pulling it wrong? I am doing
```shell
 git fetch  upstream pull/765/head:declarative-navigation
git checkout declarative-navigation
```

MV-GH commented:
Also the about page is even more wrong than before see #725, but i dont think its related to this pr. I think there might be some issues with our resource bundle

![image](https://github.com/dessalines/jerboa/assets/67873169/02a23714-bd4d-4383-a448-d13143bf4604)

nahwneeth commented:
> Found some issues though, clicking on @ linking to subs causes a crash (community@instance)

I've fixed deep links. The community activity needs some refactor and I've added a TODO for it.

> Also the about page is even more wrong than before see https://github.com/dessalines/jerboa/issues/725, but i dont think its related to this pr

And yeah, the about activity issue isn't from this PR.

MV-GH commented:
I can confirm that it is fixed now.

marcellogalhardo commented on app/src/main/java/com/jerboa/ui/components/common/InitializeRoute.kt:
```diff
@@ -0,0 +1,26 @@
+import kotlinx.coroutines.CoroutineScope
+
+@Composable
+fun InitializeRoute(initializationBlock: suspend CoroutineScope.() -> Unit) {
```
I have concerns about the `initializeRoute` function. If a Process Death happens, `initializationBlock` won't be re-called because you used `rememberSaveable` for the `isInitialized: Boolean` but the `ViewModel` will be recreated. It seems that could place the `ViewModel` in an inconsistent state.

nahwneeth replied:
Did not have process death in mind while implementing this. Looks like the isIntialized variable is better stored in the viewModel itself since that is what is currently focusing on here.

But again the dependencies will not be stored which is required for initializing the route. In fact any data in the view model is vulnerable unless we store it in the the saveStateHandle or database ig. However, callbacks cannot be stored anywhere.

One approach we could take is to check if the dependencies exist in the dependency container and redirect to home otherwise. We could prevent crashes due to inconsitent state this way and the UX would also be good. The data in acitivities like CreatePostActivity can be persisted and reloaded when the user navigates back to this activity from the home activity.

What do u think?

nahwneeth commented on app/src/main/java/com/jerboa/nav/DefaultTransitions.kt:
```diff
@@ -0,0 +1,9 @@
+import androidx.compose.animation.slideInHorizontally
+import androidx.compose.animation.slideOutHorizontally
+
+val enterTransition = slideInHorizontally { it }
```
Animation was a few lines, so included it here. Agree this should have been a different PR.

nahwneeth commented on app/src/main/java/com/jerboa/nav/NavControllerWrapper.kt:
```diff
@@ -0,0 +1,28 @@
+abstract class NavControllerWrapper {
+    protected abstract val navController: NavController
+
+    fun canPop() = navController.previousBackStackEntry != null
```
So initially, I had a Flow in there and the back button would listen to the nav stack changes and show the appropriate icon. I kinda later realized I could just remember that value.

Also, none of the routes get the navController as is. They will be wrapped and hence this makes it easier to pass navigation stuff that's common to all the routes (like canPop here).

nahwneeth commented on app/src/main/java/com/jerboa/nav/NavControllerWrapper.kt:
```diff
@@ -0,0 +1,28 @@
+// https://stackoverflow.com/a/69533584/13390651
+@OptIn(ExperimentalLayoutApi::class)
+@Composable
+fun PaddingValues.bottomIfKeyboardNotOpen(): State<Dp> {
```
Will move this to a different file.

nahwneeth replied:
One irritating thing about this is that IsImeVisible has an initial value of true which add some padding for a few milliseconds when the app is opened.

nahwneeth commented on app/src/main/java/com/jerboa/nav/NavigateImplementations.kt:
```diff
@@ -0,0 +1,17 @@
+    private val navigateToDestination: () -> Unit,
+) {
+    fun navigate(dependencies: D) {
+        container.dependencies = dependencies
```
Honestly, did not think about process death when implementing this. 

The intention:

Take a look at the PostEditViewModel in the main branch:
https://github.com/dessalines/jerboa/blob/30aa860bf40e2e7b522e9abb590ae41223085cd6/app/src/main/java/com/jerboa/ui/components/post/edit/PostEditViewModel.kt#L53-L59

Since the post can be edited from the home (or feed) activity, post activity, community activity & a person's profile activity, all these view models have to be passed to the PostEditViewModel to update the data. Usually after editing a post, you go back to any one of the these activities and it's good UX to show the updated data (for confirmation etc). Also the update logic should ideally not be in the post edit view model.

My intention was to invert this:
With this PR one can pass a callback for post edit and any activity can update itself accordingly.

In home/feed activity:
https://github.com/nahwneeth/jerboa/blob/bcefa3a67e8d81ac4398b70b2afa393dafa04028/app/src/main/java/com/jerboa/ui/components/home/FeedActivity.kt#L320-L327
```kotlin
onEditPostClick = { postView ->
    navController.toPostEdit.navigate(
        PostEditDependencies(
            postView = postView,
            onPostEdit = homeViewModel::updatePost,
        ),
    )
},
```

In post activity:
https://github.com/nahwneeth/jerboa/blob/bcefa3a67e8d81ac4398b70b2afa393dafa04028/app/src/main/java/com/jerboa/ui/components/post/PostActivity.kt#L254-L261
```
onEditPostClick = { pv ->
    navController.toPostEdit.navigate(
        PostEditDependencies(
            postView = pv,
            onPostEdit = postViewModel::updatePost,
        ),
    )
}
```

Person profile activity:
https://github.com/nahwneeth/jerboa/blob/bcefa3a67e8d81ac4398b70b2afa393dafa04028/app/src/main/java/com/jerboa/ui/components/person/PersonProfileActivity.kt#L381-L388
```
onEditPostClick = { pv ->
    navController.toPostEdit.navigate(
        PostEditDependencies(
            postView = pv,
            onPostEdit = personProfileViewModel::updatePost,
        ),
    )
},
```

I hope u get the idea. I tried a few things and I felt this was the best way to pass dependencies/callbacks. Of course, there might be a better way to do this which I'm not aware of. Open to suggestions and ideas.

The second intention is to scope the view models to the routes:
Right now these view models are declared globally and they are reused. This has resulted in stale data and the workaround currently being used to reload the view models when navigating back which is leading to other issues like the data getting reset, scroll positions not being retained and filters getting reset. These global view models also, if i'm not wrong, will not survive process death.

nahwneeth commented on app/src/main/java/com/jerboa/nav/DependencyContainer.kt:
```diff
@@ -0,0 +1,26 @@
+}
+
+@Composable
+fun <D : ViewModel> dependencyContainer(): DependencyContainer<D> {
```
Yep, the gist is instead of reusing global view models for route functionality, use global view models as proxy to move just the dependencies required for initializing the route (and it's respective viewmodel).

nahwneeth commented:
@marcellogalhardo 

I looked into the feasibility of persisting view models. I've opened an issue #810 to make the autogen types parcelable or serializable.

For now, I'll try serializing the types manually with Gson.

Moving this to draft.

nahwneeth commented:
Closing this in favor of #817 and #855
